### PR TITLE
Feature/split dataset api tests

### PR DESCRIPTION
**Pull request by nshumoogum**
### What

Split out dataset api tests into web and publishing to be run against web and publishing subnets respectively.

### How to review

Check tests pass against `feature/observation-spec` branch on `dp-dataset-api`, remember to run `make acceptance-web` or `make acceptance-publishing` to start service in web or publishing respectively. You will also need `dp-auth-api-stub` running for publishing tests to pass.

### Who can review

Team B

### CODE REVIEW

File: publishing/datasetAPI/put_instance_test.go [hunk 1]
```diff
@@ -59,9 +62,8 @@ func TestSuccessfullyPutInstance(t *testing.T) {
 
 		Convey("and is updated to a state of `completed`", func() {
 
-			datasetAPI.PUT("/instances/{instance_id}", instanceID).
-				WithHeader(florenceTokenName, florenceToken).WithBytes([]byte(validPUTCompletedInstanceJSON)).
-				Expect().Status(http.StatusOK)
+			datasetAPI.PUT("/instances/{instance_id}", instanceID).WithHeader(florenceTokenName, florenceToken).
```
Comment (eldeal) [inline publishing/datasetAPI/put_instance_test.go]: It seems inconsistent that there's a few cases in this file which are only across 2 lines instead of 4/5? Some of the lines are longer than ones that have been split out in other cases

File: publishing/datasetAPI/put_version_test.go [hunk 2]
```diff
@@ -440,19 +464,10 @@ func TestFailureToUpdateVersion(t *testing.T) {
...
 
 				datasetAPI.PUT("/datasets/{id}/editions/{edition}/versions/{version}", datasetID, edition, version).
-					WithHeader(florenceTokenName, florenceToken).WithBytes([]byte(`{"state": "associated"}`)).
-					Expect().Status(http.StatusBadRequest).Body().Contains("Missing collection_id for association between version and a collection")
-
-			})
-		})
-
-		// test for bad request when publishing version (Missing mandatory fields)
-		Convey("When an authorised PUT request is made to update version resource to a state of published", func() {
```
Comment (eldeal) [inline publishing/datasetAPI/put_version_test.go]: It feels like we probably still need some test to assert that a document can't go straight from edition confirmed to publish without meeting certain prereqs (at least the same as the associated) - though I haven't looked to see if one is buried somewhere else in the file admittedly

File: web/datasetAPI/get_datasets_test.go [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+		Value:      datasetID,
+		Update:     ValidPublishedWithUpdatesDatasetData(datasetID),
+	}
+
+	unpublishedDatasetDoc := &mongo.Doc{
+		Database:   cfg.MongoDB,
+		Collection: "datasets",
+		Key:        "_id",
+		Value:      "133",
+		Update:     validAssociatedDatasetData(datasetID),
```
Comment (eldeal) [inline web/datasetAPI/get_datasets_test.go]: Should the update be initialized with the actual dataset ID for this document - in this case 133?

File: web/datasetAPI/get_datasets_test.go [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+
+				response.Value("items").Array().Element(0).Object().Value("id").NotNull()
+
+				for i := 0; i < len(response.Value("items").Array().Iter()); i++ {
+					// User should NOT be able to see unpublished dataset in response
+					response.Value("items").Array().Element(i).Object().Value("id").NotEqual("133")
+
+					if response.Value("items").Array().Element(i).Object().Value("id").String().Raw() == datasetID {
+						// check the published test dataset document has the expected returned fields and values
+						response.Value("items").Array().Element(i).Object().Value("id").Equal(datasetID)
```
Comment (eldeal) [inline web/datasetAPI/get_datasets_test.go]: isnt this functionally the same check as the if statement? probably dont need both

File: web/datasetAPI/get_datasets_test.go [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+						checkDatasetResponse(datasetID, response.Value("items").Array().Element(i).Object())
+					}
+
+					if response.Value("items").Array().Element(i).Object().Value("id").String().Raw() == "133" {
```
Comment (eldeal) [inline web/datasetAPI/get_datasets_test.go]: again, this feels like the same assertion from further up this test, seems redundant

File: web/datasetAPI/get_edition_test.go [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+
+	if err := mongo.Setup(docs...); err != nil {
+		log.ErrorC("Was unable to run test", err, nil)
+		os.Exit(1)
+	}
+
+	datasetAPI := httpexpect.New(t, cfg.DatasetAPIURL)
+
+	Convey("Given a published edition of a dataset", t, func() {
+		Convey("When a GET requestis made to retrieve the edition", func() {
```
Comment (eldeal) [inline web/datasetAPI/get_edition_test.go]: space in requestis please :)

File: web/datasetAPI/get_editions_test.go [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+		Key:        "_id",
+		Value:      datasetID,
+		Update:     ValidPublishedWithUpdatesDatasetData(datasetID),
+	}
+
+	unpublishedEdition := &mongo.Doc{
+		Database:   cfg.MongoDB,
+		Collection: "editions",
+		Key:        "_id",
+		Value:      "466",
```
Comment (eldeal) [inline web/datasetAPI/get_editions_test.go]: This value isnt passed in to the update - should it be?

File: web/datasetAPI/get_version_test.go [hunk 8]
```diff
@@ -0,0 +1,270 @@
...
+		})
+	})
+
+	if err := mongo.Teardown(publishedDataset, publishedEdition); err != nil {
+		log.ErrorC("Was unable to run test", err, nil)
+		os.Exit(1)
+	}
+}
+
+func setupPublishedAndUnpublishedVersions(datasetID, editionID, edition, instanceID string) ([]*mongo.Doc, error) {
```
Comment (eldeal) [inline web/datasetAPI/get_version_test.go]: this function name seems misleading - this only setups published documents

**Review by eldeal — CHANGES_REQUESTED** [review]
The readmes mention a make target but I can't find any Makefiles?

File: publishing/datasetAPI/put_version_test.go (hunk 2 above)
Comment (nshumoogum) [reply publishing/datasetAPI/put_version_test.go]: As discussed earlier, the `collection_id` is now removed on publishing. Validation of field could be done on earlier states of the resource. None the less I have added some tests to check other mandatory fields are set before publishing and invalid ones too.

File: web/datasetAPI/get_datasets_test.go (hunk 3 above)
Comment (nshumoogum) [reply web/datasetAPI/get_datasets_test.go]: yes, i've made this a variable to be reused and is not hardcoded to `133`

**Comment by nshumoogum** [comment]
@eldeal  re. make targets, this is referring to the application makefile. And now that everything is merged together in `feature/observation-spec` branch on the dataset API you should be able to find them there and be able to run the service on that branch in the dataset API for the acceptance tests to pass.

Don't forget to run `dp-auth-api-stub` too for publishing tests

File: publishing/datasetAPI/get_dataset_test.go [hunk 9]
```diff
@@ -17,6 +17,7 @@ import (
 func TestSuccessfullyGetADataset(t *testing.T) {
 
 	datasetID := uuid.NewV4().String()
+	unpublishedDatasteID := uuid.NewV4().String()
```
Comment (eldeal) [inline publishing/datasetAPI/get_dataset_test.go]: there is a typo in the variable name

File: publishing/datasetAPI/get_editions_test.go [hunk 10]
```diff
@@ -117,10 +107,10 @@ func TestFailureToGetListOfDatasetEditions(t *testing.T) {
 
 	Convey("Given the dataset does not exist", t, func() {
 		Convey("When a request to get editions for a dataset is made", func() {
-			Convey("Then return a status not found (404)", func() {
+			Convey("Then return a status unauthorized (404)", func() {
```
Comment (eldeal) [inline publishing/datasetAPI/get_editions_test.go]: unauthorized is not a 404 :)

File: publishing/datasetAPI/put_version_test.go [hunk 11]
```diff
@@ -436,15 +440,75 @@ func TestFailureToUpdateVersion(t *testing.T) {
 		}
 	})
 
+	// test for updating a version so it with a state of `published` but missing mandatory fields
+	Convey("Given a published dataset and edition and an unpublished version exist ...", t, func() {
```
Comment (eldeal) [inline publishing/datasetAPI/put_version_test.go]: dont need `...` in the convey statements throughout this test. also `so it with a state` does not make sense in the comment

File: publishing/datasetAPI/put_version_test.go [hunk 12]
```diff
@@ -436,15 +440,75 @@ func TestFailureToUpdateVersion(t *testing.T) {
...
+
+		Convey("... with mandatory fields missing", func() {
+			docs, err := setupResources(datasetID, editionID, edition, instanceID, 9)
+			if err != nil {
+				log.ErrorC("Was unable to setup test data", err, nil)
+				os.Exit(1)
+			}
+
+			Convey("When an authorised PUT request is made to update version resource to published", func() {
+				Convey("Then fail to update resource and return a status bad request (400) with a message `Missing collection_id for association between version and a collection`", func() {
```
Comment (eldeal) [inline publishing/datasetAPI/put_version_test.go]: it doesnt feel like the end of this message matches what is being asserted in Contains() - this could just say 'with the correct message' or it should test for both aspects of the outcome if theres genuinely two messages (same below)

File: publishing/datasetAPI/get_datasets_test.go [hunk 13]
```diff
@@ -48,32 +49,13 @@ func TestSuccessfulGetAListOfDatasets(t *testing.T) {
...
+		Convey("when the user is authorised", func() {
 
-				if response.Value("items").Array().Element(i).Object().Value("id").String().Raw() == "133" {
-					// user is not authenticated to see this item, if it is returned force failure
-					t.Log("user is not authenticated to see this item, forced test failure")
-					t.Fail()
-				}
-			}
-		})
+			expectedDatasets := make(map[string]int)
```
Comment (eldeal) [inline publishing/datasetAPI/get_datasets_test.go]: why is this a map and not just a count? the same incrementing could happen in the for loop

Comment (nshumoogum) [reply publishing/datasetAPI/get_datasets_test.go]: So firstly yes it could just be a count of int type and secondly we wouldn't want it in the for loop as some people do not use the test database to set up test data when running tests, and hence they could have more than two datasets being returned, hence we need to look for the two datasets that we are inspecting/checking. 

Also if someone does not teardown there test data between scenarios correctly this could give a false negative here.

I will change variable to an int however

**Review by eldeal — APPROVED** [review]
